### PR TITLE
Extend URI scheme for delegation portfolio links

### PR DESCRIPTION
**Pull request by rphair**
_**Non-technical visitors:** See complete draft of CIP, including proposed changes, here: **[CIP-0013.md](https://github.com/rphair/CIPs/blob/multi-pool-links/CIP-0013/CIP-0013.md)**_

The multi pool component of this originally submitted stand-alone CIP ([CIP-COSD-MultiPoolStakeURIscheme.md](https://github.com/rphair/CIPs/blob/uri-delegation-portfolio/CIP-COSD-MultiPoolStakeURIscheme/CIP-COSD-MultiPoolStakeURIscheme.md)) was split off as a result of https://github.com/cardano-foundation/CIPs/pull/25#issuecomment-727620107 - with an agreement to add it back in after officially merging a standard for single pool links into an active CIP (which later became CIP-0013).  Some notes & ideas beyond the CIP content:

As @SebastienGllmt has suggested in meetings & media there is a chance single pool links will be implemented earlier in the wallet development timeline (we hope, earlier) than the multi pool links (which we're prepared to get later).
* To allow for this I've modified the language of that section a bit which suggests wallets should generate an error if passed a multi-pool delegation URI while still supporting only single pool links.
* I believe this is easier to follow than having 2 levels of the standard (or 2 separate CIPs) each with their own relevant ABNF grammar, syntax rules, etc.
* Many SPOs would like to get started immediately on single stake pool references even if the multi-pool links & portfolios still await difficult protocol changes like "ratio staking" and complex updates to the wallet UIs.

Multi-pool details of this Specification, especially how to interpret each decimal `proportion` given to each pool, were originally published in the latest version of this CIP draft submitted 2020-09-30: https://github.com/rphair/CIPs/blob/uri-delegation-portfolio/CIP-COSD-MultiPoolStakeURIscheme/CIP-COSD-MultiPoolStakeURIscheme.md#specificationhttps://github.com/cardano-foundation/CIPs/pull/25 > [Specification](https://github.com/rphair/CIPs/blob/uri-delegation-portfolio/CIP-COSD-MultiPoolStakeURIscheme/CIP-COSD-MultiPoolStakeURIscheme.md#specification)

My understanding of the CIP process is that we can't use recently submitted JSON specification (https://github.com/cardano-foundation/CIPs/pull/82) as a Reference implementation, so I've added a section in Rationale briefly comparing the two schemes.  After six months of discussing this I am sure the community will need both.  Some other ideas on how JSON vs. URI delegation portfolios might evolve concurrently:

* @KtorZ @wrmarchetto The rules of interpreting proportions posted in _this_ CIP, and the discussion of integer vs. decimal type and what to do if any of the `proportion` parameters are missing, might also help with more rigorous definition of the JSON standard for delegation portfolios.
* I hope we can discuss how URIs might be considered more decentralised and more resistant to censorship than hosted files; a question which came up originally [here](https://forum.cardano.org/t/cip-stake-uri-scheme-for-pools-delegation-portfolios/40594/2) (2021-09-23) when JSON was suggested as a preferred alternative to URI delegation lists.

It's a long shot that the exporting live delegation figures as a portfolio would reveal a user's identity, but I called attention to that problem in https://github.com/cardano-foundation/CIPs/pull/25 > [Specification](https://github.com/rphair/CIPs/blob/uri-delegation-portfolio/CIP-COSD-MultiPoolStakeURIscheme/CIP-COSD-MultiPoolStakeURIscheme.md#specification) and have added it as an explicit security consideration here.  It's a minor risk, but given the amount of care taken to preserve wallet anonymity for users that want it, it would be a shame to give them an accidental means of publicly identifying themselves.

### CODE REVIEW

**Comment by rphair** [comment]
p.s. BTW I'm not sure why this is happening but all activity between here & the original PR comment has already taken place in the context of https://github.com/cardano-foundation/CIPs/pull/61.

**Comment by rphair** [comment]
These are standards related items I'm not including in the CIP itself for the sake of brevity & efficiency... if any editor thinks any of these points should be included, let me know & I'll edit it back in.

### Points for Consideration

(Motivation) If we have single pool links, we need multi pool links:

* logically, since the repeated key-value structure of URIs makes this easy to understand & parse.
* practically, since saturation limit will only decrease which makes the single pool link applicable to progressively fewer uses cases as more single pools break up into multiple pools.

https://github.com/cardano-foundation/CIPs/pull/25 > [Rationale](https://github.com/rphair/CIPs/blob/uri-delegation-portfolio/CIP-COSD-MultiPoolStakeURIscheme/CIP-COSD-MultiPoolStakeURIscheme.md#rationale) (non-technical) outlined (in September 2020) many use cases for delegation portfolios and multi-pool links.

* Today it seems the need for multi-pool links is self-evident, especially after the round of applause given to the more recent submitted https://github.com/cardano-foundation/CIPs/pull/82.
* So I'm recalling these again because the list of use cases might serve as a good user stories to determine how the wallets might handle each case of a delegation portfolio, whether implemented as URI or JSON.

(from https://github.com/cardano-foundation/CIPs/pull/25 > [Specification](https://github.com/rphair/CIPs/blob/uri-delegation-portfolio/CIP-COSD-MultiPoolStakeURIscheme/CIP-COSD-MultiPoolStakeURIscheme.md#specification)) Handling of errors and warnings should be left to the application.  Examples:

* a non-numerical proportion might assume `1` (as if the figure had been left out entirely) while also displaying a warning message & asking the user if they would like to proceed according to that assumption or abort.
* a non-existent stake pool might cause a warning, with an offer to proceed without that pool, to choose an alternative, or to abort.

The original PR had a lot of comments about URI length, with a response indicating why this is not a problem for a variety of use cases, with related factors like data precision.  So it would help to check this out, to see if any of these matters should be addressed in the CIP: https://github.com/cardano-foundation/CIPs/pull/25#issuecomment-705031567

There was never a response to [this forum comment](https://forum.cardano.org/t/cip-stake-uri-scheme-for-pools-delegation-portfolios/40594/2
) (@disassembler) about the HTTP content disposition of the JSON file and how it would help certain application scenarios to have a URI alternative (the current CIP draft simply suggests that having an alternative is helpful in these cases).

The same thread had a statement which I might summarise as:
> Not requiring a JSON file also supports decentralisation though greater anonymity and censorship resistance.

... which I'm not including since so we don't get hung up in a debate about what all these things really are.  The forum thread above suggests why the Internet at large considers URIs less identity-revealing and less easily blocked than files.

@KtorZ With respect to the CIP-0017 current draft (https://github.com/cardano-foundation/CIPs/pull/82):

* I'm not planning to document in CIP-0013 why the stake pool proportions should be decimals, rather than integers as suggested in the later CIP, because no reason has yet been given for restricting these to integers.  Until such a reason is given, the general case of decimal proportions is more logical simply because it's more flexible.
* There's no need for floating point precision but to facilitate certain use cases we will actually need decimals.  I'll be posting a comment in https://github.com/cardano-foundation/CIPs/pull/82 about why I believe this is the case.

A useful link I'm not including in Read More section (though @wrmarchetto I was very happy to find out about this site last week 😍):

* https://adafolio.com (great example of Cardano delegation portfolio uses cases but already cited in https://github.com/cardano-foundation/CIPs/pull/82)

**Comment by v-almonacid** [comment]
Hey @rphair, could you please check this proposal I just submitted for comments? https://forum.cardano.org/t/cip-generalized-cardano-urls/57464
Essentially, this would mean that this PR should just be reformulated as a separate CIP (well, if the proposal makes sense for the rest of the community)

**Comment by rphair** [comment]
thanks @v-almonacid ... breaking it up that way would be OK with me because the writing regarding stake pool extensions has been generally distinct from what you and @SebastienGllmt had written already.  And I can see that there are heaps of metadata URI features on the way which would make any "Cardano URI scheme" document as long as any technical paper.

I would only insist that the merge process for this PR, which Sebastien himself suggested back in October... i.e., consolidating the multi-pool links in with the single-pool links... be completed first.  Like all standards, a coherent presentation of that standard should be dated... and the effort keeps getting greater and greater on my part to keep rewriting, refactoring, consolidating and re-posting what's already been done.

As we heard in the CIP editor meetings, where I had to catch the attendees up for 5 meetings in a row about a 6-month old proposal, the resulting history has become difficult even for people already exposed to the issue to follow: and as with all standards we need to trace this issue historically.

So while everybody is still current in the discussion of multi-pool links, hopefully this will be acceptable to @crptmppt @dcoutts @KtorZ and the other editors from that meeting, and the idea & writing of multi-pool links can be considered on their own in this PR and merged before too long.  Then I'll follow the CIP team's suggestions about breaking the pool linking scheme (for the `//stake` URI authority) into its own document.

**Comment by v-almonacid** [comment]
@rphair yeah I'm totally aware of the time and work that you've put on this, and that things have taken much more time than expected, it's unfortunate. I feel like If we had a framework as the one I'm proposing before, things could have been easier.

Anyway, it's totally fine for me if this PR gets merged in CIP13. Just keep in mind that this feature may still take a bit of time to be actually implemented by wallets, so there is no rush. I'm saying this just to suggest that we should at least double check that this is 100% compatible with the new standard, if it ever gets approved.

File: CIP-0013/CIP-0013.md [hunk 1]
```diff
@@ -75,10 +79,15 @@ paymentref = cardanoaddress [ "?" amountparam ]
...
-stakepoolref = "//stake?" stakepool
+stakepoolref = "//stake" stakequery
+stakequery = ( "?" stakepoolpair) *( "&" stakepoolpair)
+stakepoolpair = stakepool [ "=" proportion]
+
 stakepool = poolhexid | poolticker
 poolhexid = 56HEXDIG
 poolticker = 3*5UNICODE
+
+proportion = *digit [ "." *digit ]
```
Comment (KtorZ) [inline CIP-0013/CIP-0013.md:90]: If anything, it'd be good to have CIP-0017 & this extension aligned; Please see [the comment on CIP-0017 about decimals](https://github.com/cardano-foundation/CIPs/pull/82#issuecomment-827468701)

Comment (rphair) [reply CIP-0013/CIP-0013.md:90]: thanks, addressed here: https://github.com/cardano-foundation/CIPs/pull/82#issuecomment-830012606

**Comment by crptmppt** [comment]
This was again discussed at the last Editor meeting (22), please refer to the [notes](https://github.com/cardano-foundation/CIPs/blob/master/BiweeklyMeetings/05-11-2021.md) to catchup on the conversation.
This PR is scheduled to be discussed further at the next [public meeting this Tuesday](https://www.crowdcast.io/e/cip-editors-meeting-23), thank you for attending as feasible to broaden/continue the conversation.

**Comment by crptmppt** [comment]
This was discussed at the last Editor meeting ([23](https://www.crowdcast.io/e/cip-editors-meeting-23)), please refer to the [notes](https://github.com/cardano-foundation/CIPs/blob/master/BiweeklyMeetings/05-25-2021.md).
An outreach is in progress towards the community/SPOs to gather preferences and open the conversation. 
@rphair is to connect through a SPO dev townhall (likely June 10th) while a broader "CIP-themed" SPO call is tentatively scheduled for June 24th. (the _next_ CIP Editors meeting after that should be Editors meeting 26 on [July 6th](https://www.crowdcast.io/e/cip-editors-meeting-26).

**Comment by rphair** [comment]
@KtorZ I will also mention the JSON delegation portfolio alternative to the URI based ones, and will be referring SPOs to https://github.com/cardano-foundation/CIPs/pull/86 as well as this one.  If you (or any other SPOs following this issue) want that Zoom link please message me directly (I don't think they want the Zoom links for those meeting to be publicly visible).  The upcoming meeting title is **SPO community call** on Thursday, 10th June at 16:00 UTC. 😎

**Comment by crptmppt** [comment]
Discussed at last Editors meeting (24) - see [notes](https://github.com/cardano-foundation/CIPs/blob/master/BiweeklyMeetings/06-08-2021.md). (conversation might be ahead of meeting notes at this point, this is a reference)

The conversation was further opened up with two SPO meeting. 
(Adding this PR as a  "Discussion" for tomorrow's CIP Editors Meeting ([25](https://www.crowdcast.io/e/cip-editors-meeting-25) - in contrast to [PR82](https://github.com/cardano-foundation/CIPs/pull/82) in the hope of consolidating)

**Comment by crptmppt** [comment]
6/29 Editors meeting ([25](https://www.crowdcast.io/e/cip-editors-meeting-25)) discussed this PR - see [notes](https://github.com/cardano-foundation/CIPs/blob/master/BiweeklyMeetings/06-29-2021.md). (conversation might be ahead of meeting notes at this point, this is a reference) 

=> PR is now in 'Last Check': it will get another look at next week's meeting ([26](https://www.crowdcast.io/e/cip-editors-meeting-26)) and should be merged in after that - consider attending the meeting if relevant to you.

**Review by crptmppt — APPROVED** [review]

**Review by KtorZ — APPROVED** [review]

**Review by SebastienGllmt — APPROVED** [review]

**Pull request merged**